File: 6-Data_structures_arrays/lists/lib/reverse_string.ex
defmodule ReverseList do
  def list(str)
      when not is_bitstring(str)
      when byte_size(str) < 2,
      do: "hmm that is not good"

  def list(str) do
    lst = String.graphemes(str)
    reverse(lst, [])
  end

  defp reverse([], acc), do: acc |> Enum.join() |> IO.puts()

  defp reverse([h | t], acc) do
    reverse(t, [h | acc])
  end

  # def list(str) do
  #   str
  #   |> String.reverse()
  #   |> IO.puts()
  # end
end

# ReverseList.list("Hello my name is Foo")

File: 6-Data_structures_arrays/reverse_string.rb
# frozen_string_literal: true

def reverse_string(str)
  return 'hmm that is not good' if str.class != String || str.length < 2

  arr = str.chars
  result = []

  arr.length.times do
    result << arr.pop
  end

  result.join
end

# def reverse_string(str)
#   str.chars.sort_by.with_index { |_, i| -i }.join
# end

str = 'hello my name is Foo'
puts reverse_string(str)
